feat: add ui server implementation

matthew-chirgwin opened the pull request:
Note: this is in progress, and is likely to be a large PR. 


- adds TS config (derived from #29)
- implements server build
- adds placeholder dockerfile to use the build
- implements server + client module (in ts)
- updates client code to be ts based
- updates linting config for ts
- updates doc around build/commands to run

Closes: #36

Signed-off-by: Matthew Chirgwin <chirmatt@uk.ibm.com>

matthew-chirgwin committed:
feat: add ui server implementation

- adds TS config (derived from #29)
- implements server build
- adds placeholder dockerfile to use the build
- implements server + adds helmet
- implements api module
- adds other modules boilerplate
- updates linting config for ts
- updates doc around build/commands to run
- update doc to reference ts files
- changes alias approach to use tsconfig paths
- dependencies added to use this

Contributes to: #36

Signed-off-by: Matthew Chirgwin <chirmatt@uk.ibm.com>

matthew-chirgwin commented on package.json:
```diff
@@ -4,20 +4,24 @@
+    "lint:js": "eslint --config ./utils/linting/eslint.js.config.js --fix --max-warnings 0 --ignore-path ./utils/linting/.eslintignore",
+    "lint:ts": "eslint --config ./utils/linting/eslint.ts.config.js --fix --max-warnings 0 --ignore-path ./utils/linting/.eslintignore",
     "lint:styling": "stylelint --config ./utils/linting/stylelint.config.js --fix --max-warnings 0",
-    "lint:format": "prettier --config ./utils/linting/prettier.config.js --write --ignore-path ./utils/linting/.prettierignore"
```
This in error only ran prettier in the utils/linting directory

matthew-chirgwin commented on package.json:
```diff
@@ -68,9 +77,12 @@
     "stylelint-config-recommended": "^3.0.0",
     "stylelint-config-sass-guidelines": "^7.0.0",
     "stylelint-config-standard": "^20.0.0",
-    "stylelint-selector-bem-pattern": "^2.1.0",
```
removed in error, fixed in next commit

dlabaj commented on package.json:
```diff
@@ -4,20 +4,24 @@
+    "start": "npm-run-all --parallel start:client start:server",
+    "start:client": "webpack-dev-server --config utils/build/webpack.dev.js",
+    "start:server": "npx nodemon --exec 'ts-node --project utils/build/tsconfig.server.json' server/main.ts",
+    "build": "webpack --config utils/build/webpack.client.prod.js && webpack --config utils/build/webpack.server.prod.js",
```
Once typescript goes in there will be some conflicts here.  We can fix it and take this instead of what's on the typescript branch for scripts.

matthew-chirgwin replied:
Happy to discuss how/where we deliver the TS support. If it is easier, I am happy for this PR to implement yout TS reccomendations

dlabaj commented on utils/build/tsconfig.client.json:
```diff
@@ -0,0 +1,8 @@
+    "module": "esnext",
+    "jsx": "react"
+  },
+  "include": ["../../client", "../../utils"]
```
you might need exclude for server directory here.

matthew-chirgwin replied:
I shall look into this

dlabaj commented on utils/build/webpack.common.js:
```diff
@@ -81,9 +84,29 @@ const withStylingModuleLoader = returnModuleRuleWithConfig(
+const withTSModuleLoader = (tsConfigFile) =>
+  returnModuleRuleWithConfig(
+    {
+      test: /\.(tsx|ts)?$/,
```
Are we going to allow js files at all if not then this is fine.

matthew-chirgwin replied:
This is configuration for the `ts-loader`, which included in the dev/prod configs for webpack as needed. atm both JS and TS loaders are included - I will see if we can go solely with one over another

dlabaj commented on utils/build/webpack.client.prod.js:
```diff
@@ -29,17 +30,23 @@ const {
           loader: MiniCssExtractPlugin.loader,
         },
       ]),
+      withTSModuleLoader('tsconfig.client.json'),
```
We can get rid of JSModuleLoader and add js files to the TSModuleLoader.  TypeScript can handle the javascript files as well.

dlabaj commented on utils/build/webpack.common.js:
```diff
@@ -81,9 +84,29 @@ const withStylingModuleLoader = returnModuleRuleWithConfig(
+const withTSModuleLoader = (tsConfigFile) =>
+  returnModuleRuleWithConfig(
+    {
+      test: /\.(tsx|ts)?$/,
```
Can change this to `/\.(js|tsx|ts)?$/` and then we can remove babel-loader and withJSModuleLoader

dlabaj reviewed (changes requested):
Just one minor change.  Other then that looks good.

matthew-chirgwin committed:
feat: complete server implementation

- allow config files to be js or json
- generate certs for development w/ helpers
- add http/https server support

Contributes to: #36

Signed-off-by: Matthew Chirgwin <chirmatt@uk.ibm.com>

matthew-chirgwin committed:
refactor: require config logic

- change require logic to account for webpack
- fixes issues around e2e tests not running
- changes test script names to reflect what the are running (jest wont
be just unit tests)

Contributes to: #36

Signed-off-by: Matthew Chirgwin <chirmatt@uk.ibm.com>

matthew-chirgwin committed:
feat: add tests for server

- adds tests for server
- files at the root level of the server not tested - ie the server
entry code and filesystem pieces. Looking into options
- turns off coverage collection for root server directory, and
placeholder index.ts in client bootstrap
- fixes few test config issues found

Contributes to: #36

Signed-off-by: Matthew Chirgwin <chirmatt@uk.ibm.com>

matthew-chirgwin commented:
@dlabaj @nictownsend / @JQrdan now ready for review - I am looking into how we could test the `main.ts` file at the root of the server - I have something which works, but fails to collect coverage. The tests could be included if desired, but have no coverage captured. Thoughts?

JQrdan commented on linting/license-check-and-add.config.json:
```diff
@@ -1,8 +1,17 @@
+    ".git/**",
+    "./utils/headers",
+    "**.md",
+    "**.pptx",
```
we have powerpoint slides??

matthew-chirgwin replied:
We do! For some of the images in the Architecture document. I made them in Powerpoint, and checked in the source file.

JQrdan commented on linting/stylelint.config.js:
```diff
@@ -15,15 +15,4 @@ module.exports = {
   extends: rulesets.concat(['stylelint-config-prettier']),
   // ignore built files
   ignoreFiles: ['**/dist'],
-  plugins: ['stylelint-selector-bem-pattern'],
-  rules: {
-    'plugin/selector-bem-pattern': {
-      componentName: '[A-Z]+',
-      componentSelectors: {
-        initial: '^\\.{componentName}(?:-[a-z]+)?$',
-        combined: '^\\.combined-{componentName}-[a-z]+$',
-      },
-      utilitySelectors: '^\\.util-[a-z]+$',
-    },
-  },
```
Why get rid of this?

matthew-chirgwin replied:
Good spot! Rebase error, will revert

JQrdan commented on server/api/api.steps.ts:
```diff
@@ -0,0 +1,126 @@
+  'I get the expected proxied response',
+  stepWithWorld(async (world) => {
+    const { request, context, configuration } = world;
+    const securedConfig = configuration.proxy.transport.cert ? true : false;
+
+    const expectFailure = context.responseWillError ? true : false;
```
tertiary that returns boolean?

matthew-chirgwin replied:
`context.responseWillError` is optional - hence the exists and value check resulting in a boolean value

JQrdan commented on server/api/router.ts:
```diff
@@ -0,0 +1,48 @@
+      secure: cert ? true : false,
+    };
+
+    debug(`api proxy configuration`, JSON.stringify(proxyConfig));
```
should this debug be gone?

matthew-chirgwin replied:
Debug are intentional debug level statements. When the logging is delivered, they will only fire when that level is set

JQrdan commented on server/client/router.ts:
```diff
@@ -0,0 +1,50 @@
+      hasIndexFile,
+    } = getFiles(publicDir);
+
+    debug(
+      `Client is hosting ${totalNumberOfFiles} static files, ${protectedFiles.length} of which are protected and require authentication`
+    );
+    debug(`Client has index.html to serve? ${hasIndexFile}`);
```
same with these

matthew-chirgwin committed:
fix: add stylelint config removed in rebase

- reverts stylelint rebase configuration removal

Contributes to: #36

Signed-off-by: Matthew Chirgwin <chirmatt@uk.ibm.com>

nictownsend commented on README.md:
```diff
@@ -25,7 +25,9 @@ If you run into any issues while working in this repo, please check out [the tro
-- `npm run start` - runs the UI in development mode
+- `npm run test` - runs all tests for the client and server
+- `npm run start` - runs the UI client and server in development mode
+  - `npm run addDevCerts` - requires `openssl`, will generate certificates for development purposes for UI development
```
Can you add `openssl` to `Getting Started` or `Prerequisites`? 

Or, can we have a default self signed in repo + github action to refresh?

matthew-chirgwin replied:
I err away from checking in certs, hence having them generated/in a directory we ignore. Will add openssl to Getting started - I wanted it to be an option, rather than a requirement

nictownsend commented on build/webpack.client.dev.js:
```diff
@@ -45,14 +59,32 @@ const devSpecificConfig = {
+    proxy: [
+      {
+        context: ['**'],
+        target: `http${devEnvToUseTls ? 's' : ''}://localhost:3000`,
```
ENV_VAR for server port?

matthew-chirgwin replied:
Good idea - let me see what I can come up with (as this would be used in a few spots).

nictownsend commented on build/webpack.client.prod.js:
```diff
@@ -3,7 +3,7 @@
  */
 
-// prod specific plugins and webpack configuration
+// client production specific plugins and webpack configuration
```
Is there enough overlap for a `clientCommon`?

matthew-chirgwin replied:
I would need to look at both again - the rename/changes here were to reflect the server having a production build mode, and the fact TS now needs to be supported

nictownsend commented on build/README.md:
```diff
@@ -9,6 +9,8 @@ This directory contains code relating to the building of the UI at dev and produ
-- `webpack.prod.js` - webpack production build mode. Contains configuration to minify and compress all built output.
+- `webpack.client.dev.js` - webpack development build mode. Stands up a configured instance of `webpack-dev-server` to supplement the common config/to enable faster development.
+- `webpack.client.prod.js` - webpack production build mode for client code. Contains configuration to minify and compress all built output.
+- `webpack.server.prod.js` - webpack production build mode for server code. Contains configuration to minify and compress all built output.
```
Plus `webpack.server.dev`

matthew-chirgwin replied:
There is no `server.dev` configuration - ts-node + nodmon is used to enable dev of the server

nictownsend commented on build/dockerfile:
```diff
@@ -0,0 +1,34 @@
+
+# ---------------------------------------------------------------------------- #
+
+FROM node:lts-alpine3.12
```
Any reason you're fixing on a version of alpine opposed to `lts-alpine`?

matthew-chirgwin replied:
I looked for the first node provided `lts` tag, which happened to be this when I looked - happy to change

nictownsend commented on build/dockerfile:
```diff
@@ -0,0 +1,34 @@
+
+EXPOSE 3000
+
+ENTRYPOINT [ "node" , "/dist/server/main.js" ]
```
Can we have a `run.sh` instead of having the file addressed directly?

matthew-chirgwin replied:
I can - I did intend this to be a short lived file as this will need to be reviewed with the @strimzi/maintainers to align with their other docker files (and supporting tools like tini). Happy to remove this file if it is easier, or to make it match other docker files - whatever makes the most sense

nictownsend commented on build/webpack.common.js:
```diff
@@ -3,22 +3,27 @@
 // constants
 const UI_TITLE = 'Strimzi UI';
 const BUILD_DIR = path.resolve(__dirname, '../dist');
+const CLIENT_BUILD_DIR = path.resolve(__dirname, `${BUILD_DIR}/client`);
```
Do you need to resolve the client/server paths? `BUILD_DIR` is already absolute.

nictownsend replied:
Also, maybe a `ROOT_DIR` to avoid needing to use `path.resolve` with `../` every time?

matthew-chirgwin replied:
The resolve is there tfor saftey I think - I see no harm in them staying. Agree re Root dir, will update

nictownsend replied:
I think ROOT_DIR is absolute, so `const BUILD_DIR = path.resolve(${ROOT_DIR}, 'dist');` - you don't need to continually resolved with `__dirname` https://nodejs.org/api/path.html#path_path_resolve_paths

nictownsend commented on client/Contexts/README.md:
```diff
@@ -1,6 +1,6 @@
 # Contexts
 
-This UI makes use of [React Context](https://reactjs.org/docs/context.html) for top level state management. This allows for data to be stored and accessed by any component without needing to prop drill values to that component. In addition, depending on use case, contexts can be used as a component or as a React Hook, allowing for integration of state where ever and however it is required. This d ...
+This UI makes use of [React Context](https://reactts.org/docs/context.html) for top level state management. This allows for data to be stored and accessed by any component without needing to prop drill values to that component. In addition, depending on use case, contexts can be used as a component or as a React Hook, allowing for integration of state where ever and however it is required. This d ...
```
```suggestion
This UI makes use of [React Context](https://reactjs.org/docs/context.html) for top level state management. This allows for data to be stored and accessed by any component without needing to prop drill values to that component. In addition, depending on use case, contexts can be used as a component or as a React Hook, allowing for integration of state where ever and however it is required. This directory contains all the custom React Context implementations used across this UI.
```

Over zealous :)

matthew-chirgwin replied:
Good spot! Will revert - knew a find and replace would be to simple :)

nictownsend commented on client/Hooks/README.md:
```diff
@@ -1,6 +1,6 @@
 # Hooks
 
-This directory contains all re-usable custom [React Hooks](https://reactjs.org/docs/hooks-intro.html#motivation) implemented for use across this UI. Unlike Models, these hooks encapsulate re-usable/common logic views or models may want to utilise, such as translation capabilities.
+This directory contains all re-usable custom [React Hooks](https://reactts.org/docs/hooks-intro.html#motivation) implemented for use across this UI. Unlike Models, these hooks encapsulate re-usable/common logic views or models may want to utilise, such as translation capabilities.
```
```suggestion
This directory contains all re-usable custom [React Hooks](https://reactjs.org/docs/hooks-intro.html#motivation) implemented for use across this UI. Unlike Models, these hooks encapsulate re-usable/common logic views or models may want to utilise, such as translation capabilities.
```

nictownsend commented on client/Images/Images.d.ts:
```diff
@@ -2,6 +2,5 @@
-export const parameters = {
-  actions: { argTypesRegex: '^on[A-Z].*' },
-};
+declare module '*.png';
```
Symlink + code change?

matthew-chirgwin replied:
Will check this - looks like a rebase issue

nictownsend commented on client/tsconfig.json:
```diff
@@ -0,0 +1,21 @@
+{
```
This approach might not work for IDEs - not found a way for `vscode` to read two tsconfigs.

This might be resolved by my aliasing and refactored tsconfig in the i18n PR though.

matthew-chirgwin replied:
I will dig it out, but the approach I have here is reccomened in the vscode docs (and I believe is standard TS behaviour) - look in the current dir for a `tsconfig.json` file, and if not found, look in the parent (etc until one found). Agree the need for specific tsconfigs per area may not make sense with other changes going in - happy to change depending on what we feel will be ready/in first

matthew-chirgwin replied:
https://github.com/microsoft/vscode/issues/3772#issuecomment-193445854

nictownsend commented on package.json:
```diff
@@ -4,25 +4,34 @@
+    "lint": "npm-run-all 'lint:allfiles -- .' 'lint:js -- .' 'lint:ts -- .' 'lint:styling -- \"**/*.(s)css\"' 'lint:format -- .'",
     "lint:allfiles": "npm-run-all 'licence:add -- {1}' --",
-    "lint:js": "eslint --config ./linting/eslint.config.js --fix --max-warnings 0 --ignore-path ./linting/.eslintignore",
+    "lint:js": "eslint --config ./linting/eslint.js.config.js --fix --max-warnings 0 --ignore-path ./linting/.eslintignore",
```
In my i18n pr I created a single eslint with ts overrides section. Once this PR is in, I'll see if mine rebases nicely to support this.

matthew-chirgwin replied:
👍 While we still have `.js` files I still wanted them linted - this felt like a simple way of having best of both (although I am sure there are other/better ways of managing it)

nictownsend replied:
I meant, the PR I have uses one eslint config to lint both js and ts - so you'd just need `lint:ts` and point to the config for it to lint both ts and js.

nictownsend commented on server/api/router.ts:
```diff
@@ -0,0 +1,48 @@
+
+export const ApiModule: UIServerModule = {
+  moduleName,
+  addModule: (logGenerator, authFn, serverConfig) => {
```
Is this temporary until we decide on a log transport mechanism? Just I would have expected more like the singleton java `log = Log.getLogger('moduleName')` instead of passing it in.

matthew-chirgwin replied:
I see the logging staying as is here, but how it works underneath changing (I expect it will delegate to pino). The `logGenerator` here is the singleton - it has an entry and 'event' handler, where entry returns exit and debug etc. It has module name etc baked in. The logger here is meant to log the `addModule` function only (the same logger is in `res.locals`, and is meant to be used by callbacks)

nictownsend replied:
Not entirely comfortable with having to pass the log generator into a function but can revisit as this matures

nictownsend commented on server/client/client.feature:
```diff
@@ -0,0 +1,39 @@
+    | /protected.html       | 404           |
+    | /                     | 404           |
+
+    Scenario Outline: If assets can be served, the client module returns the appropriate <StatusCode> return code for a request of <Asset>
```
I think this makes above test redundant (`404` is tested here too)

matthew-chirgwin replied:
Will check, but think there are cases where the above test still has value

matthew-chirgwin replied:
Just checked - this covers the negative case where there are no files on the file system to serve (and confirming the controller returns a sensible default in this case, `[]`. The other 404 case is where a named file (requested by user) is not found

nictownsend commented on server/client/client.steps.ts:
```diff
@@ -0,0 +1,50 @@
+    await request.then(
+      (res) => {
+        const { status } = res;
+        //console.dir(res);
```
comment

matthew-chirgwin replied:
Good spot will remove

nictownsend commented on server/client/controller.ts:
```diff
@@ -0,0 +1,64 @@
+import { existsSync, readdirSync } from 'fs';
+
+// function to recursively get all files from a directory
+const getFilesInDirectory: (directory: string) => Array<string> = (directory) =>
```
Utility (with max depth to avoid infinite with symlink), or https://www.npmjs.com/package/recursive-readdir

matthew-chirgwin replied:
I am not sure changing this at this time gives us any additional benefit. The content it will recurse is what we build (which should be 'simple' - no symlinks, a moderate number of files), and is not public (so cannot be used as a utility. If there was need of something similar in other code than absolutley makes sense to harden this/use 3rd party.

nictownsend commented on server/client/router.ts:
```diff
@@ -0,0 +1,50 @@
+
+const moduleName = 'client';
+
+export const ClientModule: UIServerModule = {
```
Client feels like a misleading name, maybe FrontEnd?

matthew-chirgwin replied:
Names are fun :) . My logic was client as in code sent to client. Happy to go with a consensus on this/update all references

nictownsend commented on server/client/controller.ts:
```diff
@@ -0,0 +1,64 @@
+    : [];
+
+// mark a subset of files as public - this means any user can access them. These entries will be used in a regex - if the test passes, it will be considered public
+const publicFiles = [
```
const/utility file? like a `.gitignore` equivalent

matthew-chirgwin replied:
I don't know of anywhere else which would care/want this list. If you can think of one, let me know and can look at options

nictownsend commented on server/client/router.ts:
```diff
@@ -0,0 +1,50 @@
+    );
+
+    // if no match, not a file (path contains '.'), and we have an index.html file, redirect to it (ie return index so client navigation logic kicks in). Else do nothing (404 unless another module handles it)
+    hasIndexFile &&
```
You'll need to override `index` https://expressjs.com/en/5x/api.html#express.static

matthew-chirgwin replied:
Good spot!

nictownsend commented on server/config/config.feature:
```diff
@@ -0,0 +1,11 @@
+    Given a 'config_only' server configuration
+    And I run an instance of the Strimzi-UI server
+    When I make a 'get' request to '/config'
+    Then I get the expected config response
```
Can `expected` be defined somewhere? as a path to a file, or a name of an exported object?

matthew-chirgwin replied:
I expect these will change as implementation follows for them. The tests here are placeholder/validate their current behaviour/acts as boilerplate for the future changes for them

matthew-chirgwin committed:
feat: add additional tests

- adds tests to cover the bootstrap of the UI server

Contributes to: #36

Signed-off-by: Matthew Chirgwin <chirmatt@uk.ibm.com>

nictownsend commented on server/config/router.ts:
```diff
@@ -0,0 +1,22 @@
+
+export const ConfigModule: UIServerModule = {
+  moduleName,
+  addModule: (logGenerator, authFn) => {
```
Wonder if `authFn` should isntead be rest params with ability to add multiple middleware to the req?

matthew-chirgwin replied:
authFn could be an array (would need to check the typing I added). I think if we wanted to add other middlewares to requests there are other ways this should be done (in core, a 4th parameter etc)

nictownsend commented on server/log/router.ts:
```diff
@@ -0,0 +1,22 @@
+
+export const LogModule: UIServerModule = {
+  moduleName,
+  addModule: (logGenerator, authFn) => {
```
Not taking in third config param?

matthew-chirgwin replied:
provided, but not used - not all modules need to be config aware when they mount

nictownsend commented on server/core/app.ts:
```diff
@@ -0,0 +1,109 @@
+        authFunction(config.authentication),
+        config
+      );
+      log(
```
That's a lot of parameters...

matthew-chirgwin replied:
~Indeed - will remove for now - good spot!~ Edit: Is this re the number of args to the log call? If so, again, I expect this will change/be refined by the logging implementation. All values do serve a purpose - although all of them being repeatedly set could be reduced

nictownsend replied:
Yeah - the log params, refining by log implementation sounds good

nictownsend commented on server/core/app.ts:
```diff
@@ -0,0 +1,109 @@
+  );
+
+  // before all handlers, add context and request start/end handlers
+  app.all('*', (req, res, next) => {
```
Series of different middleware functions, might make it clearer what each one is responsible for?

matthew-chirgwin replied:
I did have this at one point, but it made things like sharing requestId less clear (assumed ordering of middleware). I would be incline to keep this as is, but happy to discuss

nictownsend replied:
Assuming request ID is a header, I think it's fair for each middleware that needs it to retrieve. This just seems like it's possibly doing three things - I'd prefer an `add header`, `add context`, `add logger` series of middleware functions.

nictownsend commented on server/log/log.feature:
```diff
@@ -0,0 +1,11 @@
+    Given a 'log_only' server configuration
+    And I run an instance of the Strimzi-UI server
+    When I make a 'get' request to '/log'
+    Then I get the expected log response
```
Same general concerns over `the expected`

nictownsend commented on server/main.ts:
```diff
@@ -0,0 +1,128 @@
+    'Error',
+    `Error thrown: ${err.message}`,
+    err,
+    others
```
From typing perspective, do we want to restrict it? Or are we likely to do `x.toString()` and end up with `[Object object]`?

matthew-chirgwin replied:
That will depend on the logger. Again, I wanted to 'shape' to be in place, with the logging issue to provide the implementation

nictownsend commented on server/main.ts:
```diff
@@ -0,0 +1,128 @@
+      JSON.stringify(latestConfig)
+    );
+    config = latestConfig;
+  }); // load config and update config value
```
Is this a description, or a TODO?

matthew-chirgwin replied:
description

nictownsend commented on server/mockapi/mockapi.feature:
```diff
@@ -0,0 +1,17 @@
+
+    Behaviours and capabilities provided by the mockapi module
+
+    Scenario: Returns with the expected response for a mocked api call
```
Is this for test purposes? Or just there until we have a full API module?

matthew-chirgwin replied:
It is placeholder, pending details of `strimzi-admin` (and thus what we have to mock)

nictownsend commented on server/placeholder.ts:
```diff
@@ -0,0 +1,62 @@
+/*
```
Can we rename to show purpose?

matthew-chirgwin replied:
Absolutley - any suggestions?

nictownsend commented on server/serverConfig.ts:
```diff
@@ -0,0 +1,103 @@
+    ];
+    // this is a deliberate require so we can load json/js, and have it parsed/modules evaluated
+    // eslint-disable-next-line @typescript-eslint/no-var-requires
+    const loadedConfig = requireForConfigLoad(pathToConfigFile);
```
What about an `fs` read to avoid the webpack magic?

matthew-chirgwin replied:
I had this at one point, but it made working with certificates at development time painful (with manual effort needed by a dev to run a secured server). This approach allows `.js` and `.json` files at runtime, meaning for dev we can use code to load the certs (and other values).

nictownsend commented on server/types.ts:
```diff
@@ -0,0 +1,153 @@
+/*
```
What's general best practice - should `types.ts` be per folder/barrel file? Or one at top level?

matthew-chirgwin replied:
I havent seen anything advocating one approach vs another. These types/interfaces are used across the server code. If there was a one time typed function, I added the definition ahead/as a part of the function. My starter for 10 was types per area (ie client/server), and one at the repo/util level if needed for both.

nictownsend commented on utils/tooling/generateCerts.sh:
```diff
@@ -0,0 +1,7 @@
+#!/bin/bash
+# Copyright Strimzi authors.
+# License: Apache License 2.0 (see the file LICENSE or http://apache.org/licenses/LICENSE-2.0.html).
+# generate 2 pairs of certificates and private keys for development purposes. Certificates valid for 10 years. Any valid certificate/key pair called `strimzi-ui-server` or `strimzi-ui-mock-admin` could be used for development, in present in the `generated/dev_certs/` directory
```
I'm not sure we want 10 years - only because chrome will moan. If it's not too slow, might be worth regenerating on startup?

matthew-chirgwin replied:
Interesting - I didn't have my chrome complain at me (above/beyond the standard self signed cert). As mentioned above, I wanted to keep openssl optional, hence no hard requirement on it/regeneration. Would a year be more sensible do you think?

nictownsend commented on server/main.steps.ts:
```diff
@@ -0,0 +1,300 @@
+  worldGenerator,
+} from '../test_common/jest_cucumber_support/commonTestTypes';
+
+import { devEnvToUseTls } from '../utils/tooling/secureDevUtil';
```
Is this dynamic per tests? Or something you expect to change infrequently in config?

matthew-chirgwin replied:
Dynamic per run. All this is doing is seeing if the certs generated by openssl are present. If they are, they are used. It does make the HTTPS test here moot if not, but there is a clear warning printed if so. I expect CI will always have certs available

matthew-chirgwin committed:
refactor: address review comments

- restore deleted file from rebase
- addresses review comments
- fixes e2e config issue

Contributes to: #36

Signed-off-by: Matthew Chirgwin <chirmatt@uk.ibm.com>

ajborley reviewed (changes requested):
~I'm seeing 2 test failures when running `npm run tests` in this branch:~
```
  ● strimzi-ui server › The server bootstraps as expected when it is not given any/missing configuration

    expect(received).toBe(expected) // Object.is equality

    Expected: 404
    Received: 200

      219 |         },
      220 |         (res) => {
    > 221 |           expect(res.statusCode).toBe(statusCode);
          |                                  ^
      222 |           resolve();
      223 |         }
      224 |       );

      at ClientRequest.<anonymous> (server/main.steps.ts:221:34)

  ● strimzi-ui server › The reloads module configuration/state when the configuration file changes

    expect(received).toBe(expected) // Object.is equality

    Expected: 404
    Received: 302

      219 |         },
      220 |         (res) => {
    > 221 |           expect(res.statusCode).toBe(statusCode);
          |                                  ^
      222 |           resolve();
      223 |         }
      224 |       );

      at ClientRequest.<anonymous> (server/main.steps.ts:221:34)
```

~Do you see these too? Are they expected?~ - now passing after clearing everything out.

When I run the dev server with `npm run start` I see servers start on ports 8080 (webpack-dev-server), 9080 (mockadminserver) and 3000 (Strimzi-ui server). Going to `http://localhost:3000/` fails with a 404 - is https://github.com/matthew-chirgwin/strimzi-ui/blob/36-ui-server/utils/dev_config/server.dev.config.js#L20 supposed to be set to `true` or would we not expect to exercise the UI via the server in dev mode? Should we only use the webpack-dev-server for client side dev/testing?

Additionally, when I try the `http://localhost:3000/api` endpoint, I get a 404 response, even though the `api` module and the `mockapi` modules are enabled in the Strimzi-ui and mockadminserver respectively. It looks like the express handler may be stripping the `api` part off the url, when it gets proxied to the mockadminserver:
```
Strimzi-ui server : Mon Oct 26 2020 14:50:38 GMT+0000 (Greenwich Mean Time) 182e278cb9580e33 - core - requestStart - GET of /api starting
Strimzi-ui server : Mon Oct 26 2020 14:50:38 GMT+0000 (Greenwich Mean Time) 182e278cb9580e33 - api - module enabled check - entry
Strimzi-ui server : Mon Oct 26 2020 14:50:38 GMT+0000 (Greenwich Mean Time) 182e278cb9580e33 - api - module enabled check - debug: Enabled? -  true
Strimzi-ui server : Mon Oct 26 2020 14:50:38 GMT+0000 (Greenwich Mean Time) 182e278cb9580e33 - api - module enabled check - exit -  true
Strimzi-ui server : Mon Oct 26 2020 14:50:38 GMT+0000 (Greenwich Mean Time) 182e278cb9580e33 - api - proxyStartHandler - entry
Strimzi-ui server : Mon Oct 26 2020 14:50:38 GMT+0000 (Greenwich Mean Time) 182e278cb9580e33 - api - proxyStartHandler - exit -  Starting proxy of request '/' to the backend api
mockadminserver : Mon Oct 26 2020 14:50:38 GMT+0000 (Greenwich Mean Time) 182e278cb9580e33 - core - requestStart - GET of / starting
mockadminserver : Mon Oct 26 2020 14:50:38 GMT+0000 (Greenwich Mean Time) 182e278cb9580e33 - client - module enabled check - entry
mockadminserver : Mon Oct 26 2020 14:50:38 GMT+0000 (Greenwich Mean Time) 182e278cb9580e33 - client - module enabled check - debug: Enabled? -  false
mockadminserver : Mon Oct 26 2020 14:50:38 GMT+0000 (Greenwich Mean Time) 182e278cb9580e33 - client - module enabled check - exit -  false
mockadminserver : Mon Oct 26 2020 14:50:38 GMT+0000 (Greenwich Mean Time) 182e278cb9580e33 - core - requestEnd - GET of / complete RC: 404
Strimzi-ui server : Mon Oct 26 2020 14:50:38 GMT+0000 (Greenwich Mean Time) 182e278cb9580e33 - api - proxyCompleteHandler - entry
Strimzi-ui server : Mon Oct 26 2020 14:50:38 GMT+0000 (Greenwich Mean Time) 182e278cb9580e33 - api - proxyCompleteHandler - exit -  Response from backend api for request '/' : 404 - Not Found
Strimzi-ui server : Mon Oct 26 2020 14:50:38 GMT+0000 (Greenwich Mean Time) 182e278cb9580e33 - core - requestEnd - GET of / complete RC: 404
```

matthew-chirgwin commented:
> Do you see these too? Are they expected?

Everything was passing for me when I went from draft -> full PR. Let me double check I haven't broken anything.

> When I run the dev server with npm run start I see servers start on ports 8080 (webpack-dev-server), 9080 (mockadminserver) and 3000 (Strimzi-ui server). Going to http://localhost:3000/ fails with a 404 - is https://github.com/matthew-chirgwin/strimzi-ui/blob/36-ui-server/utils/dev_config/server.dev.config.js#L20 supposed to be set to true or would we not expect to exercise the UI via the server in dev mode? Should we only use the webpack-dev-server for client side dev/testing?

When running `npm run start`, webpack dev server hosts the client, so a request to http://localhost:3000/ will 404 - as there is no (built) client to host. When run via e2e tests, the client is built and served by express as per a prod topology, so that will return index.html

> Additionally, when I try the http://localhost:3000/api endpoint, I get a 404 response, even though the api module and the mockapi modules are enabled in the Strimzi-ui and mockadminserver respectively. It looks like the express handler may be stripping the api part off the url, when it gets proxied to the mockadminserver..

Good spot! `/api` should not be removed - will fix and add a test!

matthew-chirgwin committed:
fix: make sure context root preserved on proxy

- makes sure context root can be preserved on proxy via new
configuration option
- test to cover

Contributes to: #36

Signed-off-by: Matthew Chirgwin <chirmatt@uk.ibm.com>

ajborley commented:
Tests are now passing after re-clearing out my node_modules and running `npm run clean`. Comment above updated. :)

matthew-chirgwin commented:
I have added a new configuration option for the context root, which will resolve the issue @ajborley mentioned re the proxying of requests loosing their context root. As we know what the context root will be for development I have set it up in the configuration so it will work as expected. Once Strimzi-admin is hosting an api, a further update will be needed to change the default context root value to the actual value.

matthew-chirgwin committed:
refactor: add multi report processing to action, rebase updates

- updates bundle action/workflows to support multiple report files/no
file cases
- updates commenter action to expect multiple bundles
- makes jest exit in node workflow (in case handles are left jest would
run forever)
- removes 'main' server tests. These do not run reliably in CI and
require reimplementation (to follow in #13)
- adds simple e2e test for client
- rebase updates
- change all dcokerfiles to use lts base images
(https://stackoverflow.com/q/64494474), and update of lock file

Contributes to: #35

Signed-off-by: Matthew Chirgwin <chirmatt@uk.ibm.com>

nictownsend commented on .github/actions/bundle/index.js:
```diff
@@ -5,16 +5,32 @@
+// map of nicer labels to built bundle files in the reports
 const BUNDLE_LABELS = {
   'main.bundle.js': "Strimzi UI JS Bundle",
+  'main.bundle..js.gz': "Strimzi UI JS Bundle (GZ compressed)",
```
`..` a typo?

matthew-chirgwin replied:
Good spot - will fix

nictownsend commented on server/mockapi/router.ts:
```diff
@@ -0,0 +1,31 @@
+
+export const MockApiModule: UIServerModule = {
+  moduleName,
+  addModule: (logGenerator) => {
```
Same comment re log generator, but maybe as @ajborley progresses it can be revisited

nictownsend reviewed (changes requested):
Minor comments

matthew-chirgwin commented:
@nictownsend / @dlabaj / @ajborley I have just updated the PR given the CI work delievered yesterday. Some notable changes:

- the 'main' tests for the server have been removed. They are not reliable, and fail in isolation (with no other tests running) in CI. I do not want to hold the PR up, and will look at alternatives in/as a part of #13 
- multiple quality of life changes around actions
  - make sure Jest exits
  - use LTS npm (fixing an issue where it would error/kill the PR workflow)
  - add a placeholder e2e cclient test so that check completes
  - tidy up the response from the commenter action
    - Print NA rather than Infinity in / 0 calculations
    - Support multiple bundles
    - *Note*: the comment is not appearing in this PR - `(node:1) UnhandledPromiseRejectionWarning: HttpError: Resource not accessible by integration` - suggests a permission issue with the action - would suggest a follow on issue. **Edit**: `..the action token for authentication is rejected when it contains changes of the main.workflow file, git push will exit with code 1` - https://github.community/t/github-action-resource-not-accessible-by-integration/16034 - may be ok from this PR on.
- regenerate/update of the package-lock due to rebase changes

This is a sizable change - can I please request final reviews from you all?

ajborley reviewed (approved):
Latest code changes LGTM 👍

nictownsend reviewed (approved):
LGTM
